server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java

```diff
@@ -0,0 +1,216 @@
+      }
+    }
+  }
+
+  @VisibleForTesting
+  void removeChoppedCF(final Key key, final BatchWriter batchWriter, final String tableName) {
+    Mutation delete = null;
+    try {
+      delete = new Mutation(key.getRow()).at().family(ChoppedColumnFamily.STR_NAME)
+          .qualifier(ChoppedColumnFamily.STR_NAME).delete();
+      log.warn(
+          "Deleting chopped reference from:{}. Previous split or delete may not have completed cleanly. Ref: {}",
+          tableName, delete.prettyPrint());
+      batchWriter.addMutation(delete);
+    } catch (MutationsRejectedException ex) {
+      log.warn("Failed to delete obsolete chopped CF reference for table: " + tableName + ". Ref: "
+          + delete.prettyPrint() + ". Will try to continue. Ref may need to be manually removed");
+      log.warn("Constraint violations: {}", ex.getConstraintViolationSummaries());
+      throw new IllegalStateException(
+          "Failed to delete obsolete chopped CF reference for table: " + tableName, ex);
+    }
+  }
+
+  @VisibleForTesting
+  void removeExternalCompactionCF(final Key key, final BatchWriter batchWriter,
+      final String tableName) {
+    Mutation delete = null;
+    try {
+      delete = new Mutation(key.getRow()).at().family(ExternalCompactionColumnFamily.NAME)
+          .qualifier(key.getColumnQualifier()).delete();
+      log.debug(
+          "Deleting external compaction reference from:{}. Previous compaction may not have completed. Ref: {}",
+          tableName, delete.prettyPrint());
+      batchWriter.addMutation(delete);
+    } catch (MutationsRejectedException ex) {
+      log.warn("Failed to delete obsolete external compaction CF reference for table: " + tableName
+          + ". Ref: " + delete.prettyPrint()
+          + ". Will try to continue. Ref may need to be manually removed");
+      log.warn("Constraint violations: {}", ex.getConstraintViolationSummaries());
+      throw new IllegalStateException(
+          "Failed to delete obsolete external compaction CF reference for table: " + tableName, ex);
+    }
+  }
+
+  /**
+   * Quick sanity check to see if value has been converted by checking the candidate can be parsed
+   * into a StoredTabletFile. If parsing does not throw a parsing exception. then assume it has been
+   * converted. If validate cannot parse the candidate a JSON, then it is assumed that it has not
+   * been converted. Other parsing errors will propagate as IllegalArgumentExceptions.
+   *
+   * @param candidate a possible file: reference.
+   * @return false if a valid StoredTabletFile, true if it cannot be parsed as JSON. Otherwise,
+   *         propagate an IllegalArgumentException
+   */
+  private boolean fileNeedsConversion(@NonNull final String candidate) {
+    try {
+      StoredTabletFile.validate(candidate);
+      log.trace("file reference does not need conversion: {}", candidate);
+      return false;
+    } catch (JsonSyntaxException ex) {
```

Comment:
Using exceptions for expected situations may be slow, needs more research.  In the quick reading I did on the subject the expensive part of throwing an exception is creating the stack trace, which makes sense.  If this is happening for a large number of files and its slow, then it could make upgrade slow.  I think we need to a follow on issue for this, it may not be a problem but would be good to investigate.  Could either try to benchmark it or just avoid throwing and catching exceptions as part of normal processing.

Reply:
Modified the needs conversion check in 882be1e05f8

Changed the check to look for a starting `{` and an ending `}`.  The data written to the metadata is well controlled, so it may be enough that it looks "json like" instead of trying fully to parse.  If the column qualifier data is a plain file reference (2.1) or something that looks like a json object starting with 3.1 and the no chop fence encoding then that is likely sufficient for what this is trying to do in a upgrade.